Display warning message when unloading is throttled.

basuke opened the pull request:
#### ba40879e6a926ad6dc5199e7ba737de922663836
<pre>
Display warning message when unloading is throttled.
<a href="https://bugs.webkit.org/show_bug.cgi?id=287180">https://bugs.webkit.org/show_bug.cgi?id=287180</a>
<a href="https://rdar.apple.com/144335494">rdar://144335494</a>

Reviewed by Chris Dumez.

If throttler decides not to grant the unloading, currently it is simply ignored.
Add console message to report the network usage exceed warning.

* LayoutTests/http/tests/iframe-monitor/resources/iframe--eligible--3.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/monitor-setup.js:
(async setup):
* LayoutTests/http/tests/iframe-monitor/throttler-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/throttler.html: Added.
* Source/WebCore/loader/ResourceMonitorThrottler.cpp:
(WebCore::ResourceMonitorThrottler::create):
* Source/WebCore/loader/ResourceMonitorThrottler.h:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::reportResourceMonitoringWarning):
* Source/WebCore/page/LocalFrame.h:
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.cpp:
(WKWebsiteDataStoreResetResouceMonitorThrottler):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::resourceMonitorThrottler):
(WebKit::WebsiteDataStore::resetResourceMonitorThrottlerForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::resourceMonitorThrottler): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didExceedNetworkUsageThreshold):
* Tools/TestWebKitAPI/Tests/WebCore/ResourceMonitor.cpp:
(TestWebKitAPI::ResourceMonitorTest::prepareThrottler):
(TestWebKitAPI::ResourceMonitorTest::disposeThrottler):
(TestWebKitAPI::ResourceMonitorTest::throttler):
(TestWebKitAPI::ResourceMonitorTest::tryAccess):
(TestWebKitAPI::TEST_F(ResourceMonitorTest, ThrottlerBasic)):
(TestWebKitAPI::TEST_F(ResourceMonitorTest, ThrottlerMaxHosts)):
(TestWebKitAPI::TEST_F(ResourceMonitorTest, ThrottlerLeastRecentAccessedHostWillBeRemoved)):
(TestWebKitAPI::TEST_F(ResourceMonitorTest, ThrottlerEmptyHostname)):
(TestWebKitAPI::ResouceMonitorTest::now): Deleted.
(TestWebKitAPI::ResouceMonitorTest::later): Deleted.
(TestWebKitAPI::TEST_F(ResouceMonitorTest, ThrottlerBasic)): Deleted.
(TestWebKitAPI::TEST_F(ResouceMonitorTest, ThrottlerMaxHosts)): Deleted.
(TestWebKitAPI::TEST_F(ResouceMonitorTest, ThrottlerLeastRecentAccessedHostWillBeRemoved)): Deleted.
(TestWebKitAPI::TEST_F(ResouceMonitorTest, ThrottlerEmptyHostname)): Deleted.
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::resetStateToConsistentValues):

Canonical link: <a href="https://commits.webkit.org/290028@main">https://commits.webkit.org/290028@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a7e491c201ab96337ae72dc7a605347b0f8d0d3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88757 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8281 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43221 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93724 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39515 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90808 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8668 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16465 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/68434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/26120 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91759 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6642 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80276 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/48799 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/6397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/34689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38624 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/76760 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35587 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95564 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15937 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/77295 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16193 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76133 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76575 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18852 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20965 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/19385 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/9002 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15951 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/21261 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15692 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19143 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17473 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1471,6 +1471,12 @@ void LocalFrame::showResourceMonitoringError()
 
+void LocalFrame::reportResourceMonitoringWarning()
+{
+    if (RefPtr document = this->document())
+        document->addConsoleMessage(MessageSource::ContentBlocker, MessageLevel::Warning, "Frame's network usage exceeded the limit."_s);
```
Can protectedDocument() be used here?

cdumez replied:
No, Basuke's code is correct. We don't want to call `protectedDocument()` since we already have a `RefPtr` from the null check above.

pvollan commented on Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.cpp:
```diff
@@ -782,3 +782,12 @@ void WKWebsiteDataStoreGetAllStorageAccessEntries(WKWebsiteDataStoreRef dataStor
     });
 }
+
+void WKWebsiteDataStoreResetResouceMonitorThrottler(WKWebsiteDataStoreRef dataStoreRef)
```
`WKWebsiteDataStoreResetResourceMonitorThrottler`

basuke replied:
Thanks for finding TYPO!

cdumez reviewed (approved):
LGTM but please fix the typo that Per Arne found.

basuke committed:
Display warning message when unloading is throttled.
https://bugs.webkit.org/show_bug.cgi?id=287180
rdar://144335494

Reviewed by Chris Dumez.

If throttler decides not to grant the unloading, currently it is simply ignored.
Add console message to report the network usage exceed warning.

* LayoutTests/http/tests/iframe-monitor/resources/iframe--eligible--3.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/monitor-setup.js:
(async setup):
* LayoutTests/http/tests/iframe-monitor/throttler-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/throttler.html: Added.
* Source/WebCore/loader/ResourceMonitorThrottler.cpp:
(WebCore::ResourceMonitorThrottler::create):
* Source/WebCore/loader/ResourceMonitorThrottler.h:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::reportResourceMonitoringWarning):
* Source/WebCore/page/LocalFrame.h:
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.cpp:
(WKWebsiteDataStoreResetResouceMonitorThrottler):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::resourceMonitorThrottler):
(WebKit::WebsiteDataStore::resetResourceMonitorThrottlerForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::resourceMonitorThrottler): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didExceedNetworkUsageThreshold):
* Tools/TestWebKitAPI/Tests/WebCore/ResourceMonitor.cpp:
(TestWebKitAPI::ResourceMonitorTest::prepareThrottler):
(TestWebKitAPI::ResourceMonitorTest::disposeThrottler):
(TestWebKitAPI::ResourceMonitorTest::throttler):
(TestWebKitAPI::ResourceMonitorTest::tryAccess):
(TestWebKitAPI::TEST_F(ResourceMonitorTest, ThrottlerBasic)):
(TestWebKitAPI::TEST_F(ResourceMonitorTest, ThrottlerMaxHosts)):
(TestWebKitAPI::TEST_F(ResourceMonitorTest, ThrottlerLeastRecentAccessedHostWillBeRemoved)):
(TestWebKitAPI::TEST_F(ResourceMonitorTest, ThrottlerEmptyHostname)):
(TestWebKitAPI::ResouceMonitorTest::now): Deleted.
(TestWebKitAPI::ResouceMonitorTest::later): Deleted.
(TestWebKitAPI::TEST_F(ResouceMonitorTest, ThrottlerBasic)): Deleted.
(TestWebKitAPI::TEST_F(ResouceMonitorTest, ThrottlerMaxHosts)): Deleted.
(TestWebKitAPI::TEST_F(ResouceMonitorTest, ThrottlerLeastRecentAccessedHostWillBeRemoved)): Deleted.
(TestWebKitAPI::TEST_F(ResouceMonitorTest, ThrottlerEmptyHostname)): Deleted.
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::resetStateToConsistentValues):

Canonical link: https://commits.webkit.org/290028@main
